src/materialized/src/http.rs

```diff
@@ -285,91 +218,160 @@ impl Server {
+    //   - HTTP Basic authentication
+    //   - JWT email address
+    //
+    // We verify that if any of these are present, they must match any other
+    // that is also present.
+
+    // First, extract the username from the certificate, validating that the
+    // connection matches the TLS configuration along the way.
+    let ConnectInfo(conn_details) = req.extensions().get::<ConnectInfo<ConnDetails>>().unwrap();
+    let mut user = match (tls_mode, conn_details) {
+        (None, ConnDetails::Http) => None,
+        (None, ConnDetails::Https { .. }) => unreachable!(),
+        (Some(TlsMode::Require), ConnDetails::Http) => return Err(AuthError::HttpsRequired),
+        (Some(TlsMode::Require), ConnDetails::Https { .. }) => None,
+        (Some(TlsMode::AssumeUser), ConnDetails::Http) => return Err(AuthError::HttpsRequired),
+        (Some(TlsMode::AssumeUser), ConnDetails::Https { client_cert }) => client_cert
+            .as_ref()
+            .and_then(|cert| cert.subject_name().entries_by_nid(Nid::COMMONNAME).next())
+            .and_then(|cn| cn.data().as_utf8().ok())
+            .map(|cn| Some(cn.to_string()))
+            .ok_or(AuthError::InvalidCertUserName)?,
+    };
+
+    // Then, handle Frontegg authentication if required.
+    let user = match frontegg {
+        // If no Frontegg authentication, we can use the cert's username if
+        // present, otherwise the system user.
+        None => user.unwrap_or_else(|| SYSTEM_USER.to_string()),
+        // If we require Frontegg auth, fetch credentials from the HTTP auth
+        // header. Basic auth comes with a username/password, where the password
+        // is the client+secret pair. Bearer auth is an existing JWT that must
+        // be validated. In either case, if a username was specified in the
+        // client cert, it must match that of the JWT.
+        Some(frontegg) => {
+            let token = if let Some(basic) = req.headers().typed_get::<Authorization<Basic>>() {
+                if let Some(user) = user {
+                    if basic.username() != user {
+                        return Err(AuthError::MismatchedUser(
+                            "user in client certificate did not match user specified in authorization header",
+                        ));
                     }
                 }
-            }),
-            _ => Box::pin(async move {
-                Response::builder()
-                    .status(StatusCode::NOT_FOUND)
-                    .body(Body::from(
-                        "The resource you have requested does not exist. Did you mean /metrics?",
-                    ))
-            }),
+                user = Some(basic.username().to_string());
+                frontegg
+                    .exchange_password_for_token(basic.0.password())
+                    .await?
+                    .access_token
+            } else if let Some(basic) = req.headers().typed_get::<Authorization<Bearer>>() {
```

Comment:
Nit: for clarity, perhaps this variable should be `if let Some(bearer)`?

Reply:
Oops! Definitely. Fixed.